### PR TITLE
Allow Jib to set the base JVM image based on the target Java version

### PR DESCRIPTION
**Pull request by geoand**
Based on https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Should.20.20jib.20autoconfig.20ubi.2017.20image.20for.20java.2017.20projects.3F

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/steps/CompiledJavaVersionBuildStep.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import io.quarkus.deployment.pkg.builditem.CompiledJavaVersionBuildItem;
+
+public class CompiledJavaVersionBuildStep {
+
+    /**
+     * Determines the Java version by looking up the major version of the first successfully parsed
+     * application .class file that is found
+     */
+    @BuildStep
+    public CompiledJavaVersionBuildItem compiledJavaVersion(BuildSystemTargetBuildItem buildSystemTarget) {
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/CompiledJavaVersionBuildStep.java:25]: @aloubyansky this is the best way to find the compiled output?

Comment (aloubyansky) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/CompiledJavaVersionBuildStep.java:25]: That'll do for now, I suppose.

**Review by aloubyansky — APPROVED** [review]

**Comment by aloubyansky** [comment]
Ah, it's still a draft, I approved it already :)

**Comment by geoand** [comment]
Thanks! It's in draft  because I am waiting on input about what the proper images should be

**Comment by geoand** [comment]
This is now a breaking change since the base JVM image was changed for JDK 11.

File: extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibConfig.java [hunk 2]
```diff
@@ -13,10 +13,13 @@
 public class JibConfig {
 
     /**
-     * The base image to be used when a container image is being produced for the jar build
+     * The base image to be used when a container image is being produced for the jar build.
+     *
+     * When the application is built against Java 17 or higher, {@code TODO} is used as the default.
```
Comment (cescoffier) [inline extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibConfig.java:18]: You may want to update this, and mention the 2 defaults

File: extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java [hunk 3]
```diff
@@ -86,14 +90,27 @@ public AvailableContainerImageExtensionBuildItem availability() {
...
+    private String determineBaseJvmImage(JibConfig jibConfig, CompiledJavaVersionBuildItem compiledJavaVersion) {
+        if (jibConfig.baseJvmImage.isPresent()) {
+            return jibConfig.baseJvmImage.get();
+        }
+
+        var javaVersion = compiledJavaVersion.getJavaVersion();
+        if (javaVersion.isJava17OrHigher() == CompiledJavaVersionBuildItem.JavaVersion.Status.TRUE) {
+            return JAVA_17_BASE_IMAGE;
+        }
+        return JAVA_11_BASE_IMAGE;
```
Comment (cescoffier) [inline extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java:113]: Is this one appCds friendly?

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibConfig.java (hunk 2 above)
Comment (geoand) [reply extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibConfig.java]: Good catch, fixed!

File: extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java (hunk 3 above)
Comment (geoand) [reply extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java:113]: Yup. I tried it and it worked like a charm

**Review by cescoffier — APPROVED** [review]

**Comment by gsmet** [comment]
While it is a breaking change, I wonder if we should push it to 2.6 anyway.

**Comment by geoand** [comment]
Yeah, I had the same thought. I don't see why not.

**Comment by maxandersen** [comment]
didnt we not also break it for the dockerfile scenario too ?

**Comment by geoand** [comment]
> didnt we not also break it for the dockerfile scenario too ?

Not sure, I haven't seen any docker related changes

**Comment by rsvoboda** [comment]
@geoand this change could have an entry in docs.

**Comment by geoand** [comment]
Why do you think so? I mean the configuration property is documented and we'll add an entry to the migration guide for `2.7`.

**Comment by rsvoboda** [comment]
Migration guide is fine, thanks!

**Comment by geoand** [comment]
The entry is [here](https://github.com/quarkusio/quarkus/wiki/Migration-Guide-2.7)

**Comment by rsvoboda** [comment]
Great, thanks.

**Comment by geoand** [comment]
YW!

**Pull request merged**